platform/jewel/ui/src/main/kotlin/org/jetbrains/jewel/ui/painter/hints/EmbeddedToInlineCssStyleSvgPatchHint.kt

```diff
@@ -0,0 +1,436 @@
+     * <circle style="fill:red;opacity:0.9"/>
+     * <!-- Note: inline opacity overrides class opacity -->
+     * ```
+     *
+     * @param cache Pre-built cache containing CSS class-to-rule mappings
+     */
+    private fun Element.inlineStyleDeclarations(cache: CssClassAttributesCache, parser: CssParser) {
+        val classAttributeName = "class"
+        val styleAttributeName = "style"
+
+        // Use XPath to efficiently find only elements with a class attribute
+        for (element in getElementsWithAttributeXPath(classAttributeName)) {
+            if (!element.hasAttribute(classAttributeName)) continue
+
+            val classAttribute = element.getAttribute(classAttributeName)
+            if (classAttribute.isBlank()) {
+                element.removeAttribute(classAttributeName)
+                continue
+            }
+
+            // Handle multiple classes: "st0 st9" or single class: "st0"
+            val classNames = classAttribute.trim().split(Regex("\\s+"))
+            val mergedStyleProperties = mergeStylePropertiesFromCssClasses(classNames, cache)
+
+            if (mergedStyleProperties.isEmpty()) {
+                // No styles found, just remove the class attribute
+                element.removeAttribute(classAttributeName)
+                continue
+            }
+
+            // Merge with existing inline styles (inline takes precedence)
+            val existingStyle = element.getAttribute(styleAttributeName)
+            if (existingStyle.isNotEmpty()) {
+                val inlineStyles = parser.parseInlineStyle(existingStyle)
+                // Inline styles override class styles
+                mergedStyleProperties.putAll(inlineStyles)
+            }
+
+            // Set the merged styles
+            val styleValue = mergedStyleProperties.entries.joinToString(";") { "${it.key}:${it.value}" }
+            element.setAttribute(styleAttributeName, styleValue)
+
+            // Remove class attribute after inlining
+            element.removeAttribute(classAttributeName)
+        }
+
+        // Remove all <style> elements from the document
+        removeAllStyleElements()
+    }
+
+    private fun mergeStylePropertiesFromCssClasses(
+        classNames: List<String>,
+        cache: CssClassAttributesCache,
+    ): MutableMap<String, String> {
+        // Resolve and merge styles from all classes
+        val mergedStyles = mutableMapOf<String, String>()
+        for (className in classNames) {
+            val classStyleProperties: Map<String, String>? = cache.getCssPropertiesFor(className)
+            if (classStyleProperties != null) {
+                // Later classes override earlier ones
+                mergedStyles.putAll(classStyleProperties)
+            }
+        }
+        return mergedStyles
+    }
+
+    /** Uses XPath to efficiently query elements with a specific attribute. */
+    private fun Element.getElementsWithAttributeXPath(attributeName: String): List<Element> {
```

Comment:
```suggestion
    private fun Element.getElementsWithAttribute(attributeName: String): List<Element> {
```

The "XPath" part of the function name seems unnecessary and a bit confusing, since it makes it sound like the parameter should be an XPath expression, but the usage of XPaths is just an implementation detail